### ISSUE TITLE
CastAndFree rule misidentifies most method calls as casts

### ISSUE THREAD

**Issue by fourls** [integrated-application-development/sonar-delphi#115]
### Prerequisites

- [X] This bug is in SonarDelphi, not SonarQube or my Delphi code.
- [X] This bug has not already been reported.

### SonarDelphi version

1.0.0

### Issue description

The CastAndFree rule is intended to catch cases where a cast is performed immediately prior to calling `Free` or `FreeAndNil`. Currently, it also picks up almost any routine call with an argument list as a cast and incorrectly flags them as redundant casts. This commonly occurs in code where the intention is to immediately discard the return value of a function.

This is due to this problematic code in `CastAndFreeCheck.java`:

```java
private static boolean isHardCast(ExpressionNode expr) {
  return expr instanceof PrimaryExpressionNode
      && expr.getChild(0) instanceof NameReferenceNode
      && expr.getChild(1) instanceof ArgumentListNode
      && expr.getChildren().size() < 6;
  }
```

I believe this rule predates semantic analysis, hence its inaccuracy - it could be easily rewritten to be accurate using SonarDelphi's modern semantic analysis tools (like in `AbstractCast` and descendants).

### Steps to reproduce

Scan a Delphi file including the following:

```delphi
function DoSomething(Obj: TObject): TObject;
begin
  Result := Obj;
end;

procedure Test;
var
 MyObj: TObject;
begin
  FreeAndNil(DoSomething(TObject.Create));
end;
```

**Comment by Cirras** [comment]
This should be easy to fix - thanks!
